upup/pkg/fi/cloudup/azure/verifier.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2022-08-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2022-05-01/network"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+	"k8s.io/kops/pkg/bootstrap"
+	"k8s.io/kops/pkg/nodeidentity/azure"
+	"k8s.io/kops/pkg/wellknownports"
+)
+
+type AzureVerifierOptions struct {
+}
+
+type azureVerifier struct {
+	client *client
+}
+
+var _ bootstrap.Verifier = &azureVerifier{}
+
+func NewAzureVerifier(ctx context.Context, opt *AzureVerifierOptions) (bootstrap.Verifier, error) {
+	azureClient, err := newClient()
+	if err != nil {
+		return nil, err
+	}
+
+	return &azureVerifier{
+		client: azureClient,
+	}, nil
+}
+
+func (a azureVerifier) VerifyToken(ctx context.Context, rawRequest *http.Request, token string, body []byte, useInstanceIDForNodeName bool) (*bootstrap.VerifyResult, error) {
```

Comment:
This doesn't validate that the VM should actually be allowed to join the cluster, it seems to accept a token from any VM in the resource group. For comparison, the AWS verifier checks that the assumed role in the GetCallerIdentity response matches an IAM role used by any of the cluster's instance groups.

Can we do something similar here?

Reply:
True. It checks at least to be in same resource group. Will check to see if I can lock down things more here.

Reply:
There may be a more secure solution by using the attested data, though will be something for the future.
For now I added the VM ID, which should be pretty hard to guess without access to the resource group.
https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux#attested-data

Reply:
Can we verify the VM's VMSS is of a known kOps InstanceGroup?

Reply:
That should be checked further along in the process. This is why we set `result.InstanceGroupName`.

Reply:
It may be safer to take the IG name from the VMSS name, instead of the tag. WDYT?